### DISCUSSION TITLE
Non-nullable parameters syntax rationale

### DISCUSSION THREAD

**Discussion by heyajulia** [dotnet/csharplang#6006] [Q&A] [answered]
With C# 11, we finally get [a syntax for non-nullable parameters](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-11#simplified-parameter-null-checks) (is that the right term?) that lets us do away with explicit `null` checks and instead lets us use a sleek syntax, `Type paramName!!`.

My question is about those two exclamation marks.

Why isn't it `Type paramName!` (one exclamation mark) or `Type! paramName` (which mirrors `Type? paramName`) or something else entirely?

**Comment by bernd5** [comment, marked as answer]
A non nullable parameter is simply `Type paramName` in nullable aware code. With `Type paramName!!` you get a non nullable parameter with an explicit runtime check (a potential nullable violation causes just a warning which can be ignored or suppressed).
This can be used for example for your public API which wants to enforce non null values for certain parameters. If your API might be used in a non-nullable aware environment you can avoid issues based on contract violations with this feature.

**Reply by heyajulia** [reply]
Yes, I know that. I'm just curious why that specific syntax was chosen.

**Reply by bernd5** [reply]
It is not `Type! paramName` because this feature does not affect the type - the type remains untouched.
`Type paramName!` would be an option

**Reply by heyajulia** [reply]
Right! That makes sense. I don't think there's some fundamental law of the universe that would prevent `Type!` from being a thing, but it would make the language somewhat inconsistent. Thank you!

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-04-06.md#parameter-null-checking

**Reply by heyajulia** [reply]
Thanks, that was very interesting!

**Comment by RikkiGibson** [comment]
> we finally get [a syntax for non-nullable parameters](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-11#simplified-parameter-null-checks) (is that the right term?)

I would use the term *null-checked parameters* instead.
